### PR TITLE
Feat: Support trait with custom resource be dispatched to hubcluster

### PR DESCRIPTION
**Pull request by fourierr**
Signed-off-by: Xiangbo Ma <maxiangboo@cmbchina.com>

### Description of your changes
Support trait with custom resource be dispatched to hubcluster
Fixes https://github.com/oam-dev/kubevela/issues/3754

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 1]
```diff
@@ -193,7 +193,7 @@ func (h *AppHandler) applyComponentFunc(appParser *appfile.Parser, appRev *v1bet
 		}
 		wl.Ctx.SetCtx(ctx)
 
-		readyWorkload, readyTraits, err := renderComponentsAndTraits(h.r.Client, manifest, appRev, overrideNamespace, env)
+		readyWorkload, readyTraits, err := renderComponentsAndTraits(h.r.Client, manifest, appRev, clusterName, overrideNamespace, env)
```
[the remark this answers is not in the document]
Comment (fourierr) [reply pkg/controller/core.oam.dev/v1alpha2/application/generator.go:196]: @sonatype-lift ignore

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 2]
```diff
@@ -169,7 +169,7 @@ func (h *AppHandler) renderComponentFunc(appParser *appfile.Parser, appRev *v1be
 		if err != nil {
 			return nil, nil, err
 		}
-		return renderComponentsAndTraits(h.r.Client, manifest, appRev, overrideNamespace, env)
+		return renderComponentsAndTraits(h.r.Client, manifest, appRev, clusterName, overrideNamespace, env)
```
[the remark this answers is not in the document]
Comment (fourierr) [reply pkg/controller/core.oam.dev/v1alpha2/application/generator.go:172]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (fourierr) [reply pkg/controller/core.oam.dev/v1alpha2/application/generator.go:172]: @sonatype-lift ignore

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fourierr) [reply pkg/controller/core.oam.dev/v1alpha2/application/generator.go:196]: @sonatype-lift ignore

File: pkg/controller/core.oam.dev/v1alpha2/application/generator.go [hunk 3]
```diff
@@ -224,6 +224,20 @@ func (h *AppHandler) applyComponentFunc(appParser *appfile.Parser, appRev *v1bet
 	}
 }
 
+func checkHubClusterOnlyTraits(appRev *v1beta1.ApplicationRevision, readyTraits []*unstructured.Unstructured) []*unstructured.Unstructured {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/generator.go:227]:
```suggestion
func overrideClusterForControlplaneOnlyTraits(appRev *v1beta1.ApplicationRevision, readyTraits []*unstructured.Unstructured) []*unstructured.Unstructured {
```

**Review by wonderflow — APPROVED** [review]
generally LGTM after nits

**Comment by fourierr** [comment]
I merged the master branch on this pr,so make a new pr(https://github.com/oam-dev/kubevela/pull/3774), and close this.

**Pull request closed without merging**